dataservice/api/diagnosis/models.py

```diff
@@ -66,3 +68,48 @@ class Diagnosis(db.Model, Base):
                                db.ForeignKey('participant.kf_id'),
                                doc='the participant who was diagnosed',
                                nullable=False)
+    biospecimen_id = db.Column(KfId(),
+                               db.ForeignKey('biospecimen.kf_id'),
+                               doc='the biospecimen which recieved the '
+                               'diagnosis',
+                               nullable=True)
+
+    @validates('biospecimen_id')
+    def validate_biospecimen(self, key, value):
+        """
+        Ensure that both the diagnosis and biospecimen
+        (referred to by biospecimen_id) have the same participant
+
+        If this is not the case then raise DatabaseValidationError
+        """
+        from dataservice.api.biospecimen.models import Biospecimen
+        from dataservice.api.errors import DatabaseValidationError
+
+        if ((self.kf_id is None) or
```

Comment:
Returning the value if kf_id is None means that we can avoid this constraint when creating a new Diagnosis?

Reply:
Eh yea, this doesn't look right. The original thought was to do a check for self.kf_id and self.participant_id being null and circumvent the constraint logic if True. During testing, I noticed what I believe to be a timing issue. It seems that when diagnosis is initialized these two fields are set to null and the DatabaseValidationError exception gets thrown. I think what i want here is to check it `((self.kf_id == None and self.participant_id == None) or value == None)` and if True, return None.

Reply:
`kf_id` will be `None` until the session is flushed. `participant_id` should  not be `None`, unless there is an issue with the tests, though?